connectors/byoei.py

```diff
@@ -130,11 +138,56 @@ async def _batch_bulk(self, operations):
                         if "error" in data:
                             logger.error(f"operation {op} failed, {data['error']}")
                             raise Exception(data["error"]["reason"])
+
+            self._populate_stats(operations, res)
```

Comment:
Potential problem with design: Bulker is supposed to be sharable between different jobs, IMO, because it's not attached to specific index.

You can provide docs that go to different indices to bulker and it would not care, and upload them all. So one Bulker can be used to write into 3 indices, so stats from operation will combine these 3 processes. Moreover, if 2 jobs (hypothetically) write into same index, the stats will be wrong.

We can group by index name of course, but we might also revisit the design of these classes to force them to have 1 instance of Bulker per index. cc @tarekziade is this code working this way (allowing 1 Bulker usage per multiple indices) to limit Memory consumption?

Reply:
>Bulker is supposed to be sharable between different jobs

100% agree, but not now. When you call `async_bulk`, it creates a new `Fetcher`, a new `Bulker`, and a new `MemQueue`. 

I had proposed that we should have one long-running `BulkerService`, which takes a `MemQueue` object as a param, and keep fetching docs from the queue and bulk to ES. In that design, each job execution will just create a Fetcher, which fetches data from the remote source and push it to the queue. And there's no need to send `END_DOCS`.

The only drawback is that it's hard (if not impossible) to get 100% accurate **indexed** stats.

This proposal is transparent to customers so we could leave it after 8.8 GA.